### ISSUE TITLE
Upload speeds are significantly slower with most recent changes in main

### ISSUE THREAD

**Issue by rpendleton** [abiosoft/colima#232]
I've been experiencing the well-known networking issues where network connectivity seems to stop working after heavy load. I noticed changes for networking were pushed to main, so I decided to test them out to see if things have improved.

I'm not sure if there's more that I need to do on my side to configure things correctly, but after installing with `brew install --HEAD colima` and then recreating and restarting the Colima VM, I'm seeing significantly worse networking performance now. Especially when it comes to upload speed.

Speed tests run outside of Docker on the host computer:
```
host computer outside of docker (wifi)
  Your download speed is 181 Mbps
  Your upload speed is 152 Mbps
  Your ping is 5 ms

host computer outside of docker (ethernet)
  Your download speed is 528 Mbps
  Your upload speed is 554 Mbps
  Your ping is 3 ms
```

Speed tests run on Colima before the updates in main:
```
colima 5a4a704 (before update, wifi)
  Your download speed is 170 Mbps
  Your upload speed is 113 Mbps
  Your ping is 5 ms

colima 5a4a704 (before update, ethernet)
  Your download speed is 559 Mbps
  Your upload speed is 337 Mbps
  Your ping is 3 ms
```

Speed tests run on Colima after the updates in main:
```
colima 5e2e413 (after update, wifi)
  Your download speed is 101 Mbps
  Your upload speed is 2 Mbps
  Your ping is 5 ms

colima 5e2e413 (after update, ethernet)
  Your download speed is 187 Mbps
  Your upload speed is 14 Mbps
  Your ping is 3 ms
```

So to summarize, after the update:

On Wi-Fi:
* Download speeds have dropped from 170 Mbps to 101 Mbps
* Upload speeds have dropped from 113 Mbps to 2 Mbps

On Ethernet:
* Download speeds have dropped from 559 Mbps to 187 Mbps
* Upload speeds have dropped from 337 Mbps to 14 Mbps

Is this slow down expected? And if so, is there some way I can opt in to using the previous networking once this is released?

**Comment by abiosoft** [comment]
Thanks for pointing this out. The download speed drop is still bearable but wow, the upload sleep is a massive drop. 
If that is indeed the case, I guess it makes more sense for users to be able to toggle it.

Any idea if the network is now more stable, even if slower?
Also, how are you running the speed tests?

**Comment by jandubois** [comment]
@abiosoft Unfortunately this is expected (even though quite surprising to me). It was even worse before; see my timings in https://github.com/virtualsquare/vde-2/pull/35.

I've tested this also with PTP mode, hoping it would improve things noticeably; unfortunately it didn't.

It is hard to believe that the slowdown comes from vmnet, but somebody needs to spend some serious time investigating this to say for sure.

**Comment by rpendleton** [comment]
@abiosoft I haven’t tested things thoroughly enough to be confident yet, so I’d take this with a grain of salt. But…

I can say in the first 5 minutes of testing the latest changes, I actually got the same failures pushing an image that I’d seen with the original networking code. Basically, a layer or two pushes fine, and then docker-compose starts showing “retrying”. And like in the previous cases, restarting the VM fixed it.

So for me, the change really doesn't seem to improve things at all. It’s possible (but unlikely) that this was just an unlucky coincidence though, so I’d like to see what others observe.

**Comment by abiosoft** [comment]
A flag to toggle between slirp and vmnet as the default route should suffice for now, vmnet will still be used for an accessible IP address regardless.

> So for me, the change really doesn't seem to improve things at all. It’s possible (but unlikely) that this was just an unlucky coincidence though, so I’d like to see what others observe.

Yeah, I just realised that my focus was solely on downloads which is now more stable e.g. the following test now runs successfully. As for uploads, I noticed a 90% drop in speed but I have not done extensive tests.

```
docker run --rm jordi/ab -c 100 -n 1000 https://files.pythonhosted.org/packages/d9/5a/e7c31adbe875f2abbb91bd84cf2dc52d792b5a01506781dbcf25c91daf11/six-1.16.0-py2.py3-none-any.whl
```

**Comment by jcdickinson** [comment]
What about using gvproxy? https://github.com/containers/gvisor-tap-vsock

**Comment by abiosoft** [comment]
> What about using gvproxy? https://github.com/containers/gvisor-tap-vsock

There is a report of it being slow as well https://github.com/lima-vm/lima/issues/561#issuecomment-1014249727.

Nonetheless, it would not be as straightforward to give gvisor a try until either Lima supports it (less feasible) or Lima permits the customization of Qemu startup args (more realistic).

**Comment by abiosoft** [comment]
@jcdickinson @rpendleton are you willing to assist with testing gvproxy?

I am actually getting similar speed to the host for both uploads and downloads with gvproxy 🤓

**Comment by jcdickinson** [comment]
> @jcdickinson @rpendleton are you willing to assist with testing gvproxy?
> 
> I am actually getting similar speed to the host for both uploads and downloads with gvproxy 🤓 

Absolutely

**Comment by jcdickinson** [comment]
@abiosoft is the test you're looking for, or are there flags you need me to pass?

```
❯ colima version
colima version v0.3.4-7-gc42a673
git commit: c42a6735898fd68747682343861eb74ea683b643

runtime: docker
arch: aarch64
client: v20.10.9
server: v20.10.14

❯ colima start
INFO[0000] starting colima
INFO[0000] runtime: docker
INFO[0000] preparing network ...                         context=vm
INFO[0000] creating and starting ...                     context=vm
INFO[0043] provisioning ...                              context=docker
INFO[0043] starting ...                                  context=docker
INFO[0049] done

❯ colima ssh

colima$ wget www.google.com
Connecting to www.google.com (142.251.16.99:80)
saving to 'index.html'
index.html           100% |*************************************************| 15803  0:00:00 ETA
'index.html' saved
colima:/Users/jono/Code/colima$ cat index.html
<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="en">

colima$ sudo apk add speedtest-cli
Download: 142.95 Mbit/s
Upload: 195.90 Mbit/s

colima$ exit

❯ docker run -it --rm alpine:latest
/ # speedtest
Download: 147.04 Mbit/s
Upload: 166.26 Mbit/s

❯ docker run --rm -it -p 8081:80 strm/helloworld-http
Unable to find image 'strm/helloworld-http:latest' locally
latest: Pulling from strm/helloworld-http
85b1f47fba49: Pull complete
8f137bf105ee: Pull complete
8fb76b9403d9: Pull complete
b53610b50bff: Pull complete
Digest: sha256:bd44b0ca80c26b5eba984bf498a9c3bab0eb1c59d30d8df3cb2c073937ee4e45
Status: Downloaded newer image for strm/helloworld-http:latest
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Serving HTTP on 0.0.0.0 port 80 ...
172.17.0.1 - - [03/May/2022 17:49:54] "GET / HTTP/1.1" 200 -
172.17.0.1 - - [03/May/2022 17:49:54] code 404, message File not found
172.17.0.1 - - [03/May/2022 17:49:54] "GET /favicon.ico HTTP/1.1" 404 -
```

I also tested Portainer and it seems to be working.

**Comment by abiosoft** [comment]
@jcdickinson the default is currently gvproxy btw. It looks okay from my tests but would like to hear other experiences before concluding on it.

I am primary looking to test which provides the best performance from the network options.

There are now three options

1. `colima start --network-driver gvproxy`. using gvproxy (currently the default)
2. `colima start --network-driver slirp`. using qemu user mode network (Qemu's default and the default in the Colima v0.3.4)
3. `colima start --network-driver vmnet` using vmnet. Vmnet is still being utilised for external IP address regardless. It has slow uploads and led to the creation of this issue.

**Comment by jcdickinson** [comment]
Here are some comparative tests:

## Internet

### slirp 7 CPU

* Download: 145.54 Mbit/s
* Upload: 308.42 Mbit/s
* CPU: ~200%

### vmnet 7 CPU

* Download: 143.13 Mbit/s
* Upload: 306.86 Mbit/s
* CPU: ~200%

### gvproxy 7 CPU

* Download: 143.29 Mbit/s
* Upload: 195.13 Mbit/s
* CPU: ~200%
 
## Host <-> VM

### slirp 7 CPU

|  Interval    |       Transfer  |   Bitrate    |     Retr   |  Party
|--|--|--|--|--|
|  [5]   0.00-10.00  sec | 2.14 GBytes | 1.84 Gbits/sec  |  0  |           sender
|  [5]   0.00-10.00  sec | 2.14 GBytes | 1.84 Gbits/sec  |     |           receiver

* CPU: ~100%

### vmnet 7 CPU

|  Interval    |       Transfer  |   Bitrate    |     Retr   |  Party
|--|--|--|--|--|
|  [5]   0.00-10.00  sec | 2.08 GBytes | 1.79 Gbits/sec  |  0  |           sender
|  [5]   0.00-10.00  sec | 2.08 GBytes | 1.79 Gbits/sec  |     |           receiver

* CPU: ~75%

### gvproxy 7 CPU

|  Interval    |       Transfer  |   Bitrate    |     Retr   |  Party
|--|--|--|--|--|
| [5]   0.00-10.00  sec | 2.14 GBytes | 1.84 Gbits/sec  |  0    |         sender
| [5]   0.00-10.00  sec | 2.14 GBytes | 1.84 Gbits/sec   |      |         receiver

* CPU: ~100%

**Comment by ramunasd** [comment]
I can confirm that latest HEAD-c42673 has normal upload speeds. Before it was <50kbps
